portality/models/account.py

```diff
@@ -152,12 +149,19 @@ def prep(self):
     @property
     def api_key(self):
         if self.has_role('api'):
-            return self.data.get('api_key', None)
+            # Return the stored api key if it's there, generate one if not (there should be one present if the role is)
```

Comment:
> Anyway, there shouldn't be many opportunity for an account with the api role to be decoupled from its api_key record in data.

Yeah, I don't think there's currently a way for a key to be deleted, even admins can't do it. At worst a developer can go in and remove it by hand, but there's a lot of havoc you can wreak that way anyway.



> My thought at the time was that returning None usually means you don't have API access, and the missing key is an accident so should be fixed by generating a new one.

If we want to do that, we need to do it in the controller (/ view function / helper) that discovers the problem, and .save() the object. It isn't really possible to authenticate against a key that is stored only in memory, the user will never know what their key is even if you generate one on the fly.

I don't think even that is necessary though, if your key is not working you would go to your account and copy it from there, if it still doesn't work you'd contact DOAJ for support.

The way to remove API permissions from somebody is to remove the api role - they can keep the key. All our private routes should check for the API role.